### PR TITLE
fix(gateway): cache eviction caused by signature mismatch and scoping bugs

### PR DESCRIPTION
**Pull request by fayenix**
## Summary

Post-turn agent cache eviction is silently triggered by two independent bugs:

1. **Model normalization mismatch**: The cache key is built using the raw model string, but `_run_agent()` normalizes it (lowercase, provider prefix). When the normalized form differs from the raw string, the signature check fails and the agent is evicted — even though nothing actually changed.

2. **`was_auto_reset` and `message_id` destabilization**: These per-request fields are included in the cache signature but change between messages, causing false-positive eviction on every turn.

3. **`reset_context_note` scoping bug**: The variable is referenced from an outer scope in `_run_agent()` instead of being passed as a parameter, causing UnboundLocalError in edge cases.

4. **Diagnostic log bug**: The cache eviction debug log prints the wrong variable, making it impossible to diagnose why eviction occurred.

## Changes

- Normalize model string consistently in the cache key builder.
- Exclude `was_auto_reset` and `message_id` from the cache signature.
- Pass `reset_context_note` as an explicit parameter to `_run_agent()`.
- Fix the diagnostic log to print the actual signature values.

## Testing

- Manual: confirmed agents remain cached across turns with same model.
- Diagnostic logs now show correct before/after signatures.

## Related

Resubmission of #21299 (closed without merge — commits preserved locally, re-cherry-picked onto current main).

### CODE REVIEW

File: gateway/run.py [hunk 1]
```diff
@@ -16276,10 +16282,23 @@ def _interim_assistant_cb(text: str, *, already_streamed: bool = False) -> None:
...
                 # Config changed or first message — create fresh agent
+                _cached_for_log = cached
+                if _cached_for_log:
+                    logger.info("CACHE MISS: sig changed (old=%s new=%s)", 
+                               _cached_for_log[1][:8], _sig[:8])
+                else:
+                    logger.info("CACHE MISS: no cached agent for session %s (sig=%s)", session_key, _sig[:8])
+                # Inject auto-reset notice into ephemeral system prompt AFTER
+                # _sig computation so one-time transient notices don't contaminate
+                # the agent cache signature and cause per-turn evictions.
```
Comment (teknium1) [inline gateway/run.py:15833]: This makes the auto-reset note invisible to the cache signature, but auto-reset creates a new session under the same session_key while the signature does not include session_id. If the previous agent is still cached, this can reuse the old AIAgent across a fresh-session boundary instead of rebuilding it.

File: gateway/run.py [hunk 2]
```diff
@@ -16276,10 +16282,23 @@ def _interim_assistant_cb(text: str, *, already_streamed: bool = False) -> None:
                             except KeyError:
                                 pass
                         self._init_cached_agent_for_turn(agent, _interrupt_depth)
-                        logger.debug("Reusing cached agent for session %s", session_key)
+                        logger.info("CACHE HIT: reusing agent for session %s (sig=%s)", session_key, _sig[:8])
```
Comment (teknium1) [inline gateway/run.py:15822]: `cached` is only assigned inside the `_cache_lock and _cache is not None` branch above. Cache-disabled runners can reach this line with no local `cached`; initialize `cached = None` before the lookup block.

**Review by teknium1 — COMMENTED** [review]
Thanks for resubmitting this. The premise is real on current main, but I found two blocking issues in the patch shape.

### Problems

- `gateway/run.py:16297` moves the auto-reset note after signature computation. Current main creates a fresh session with the same `session_key` at `gateway/session.py:950-963`, and the cache signature does not include `session_id` (`gateway/run.py:12583-12600`, call at `gateway/run.py:14363-14371`). If the old cached agent hits, it can survive across the reset boundary with the old `session_id` and transcript.
- `gateway/run.py:16286` reads `cached` outside the only branch that assigns it. Cache-disabled runners exist in tests, e.g. `tests/gateway/test_proxy_mode.py:21-23`, so this can raise `UnboundLocalError` when `_agent_cache_lock` is absent.
- The PR currently has manual testing only. Existing cache tests cover generic signatures (`tests/gateway/test_agent_cache.py:30-46`) but not the model-normalization, Discord `message_id`, or auto-reset boundary cases changed here.

### Suggested changes

- Keep auto-reset as an explicit cache boundary: evict before `_run_agent`, or include a stable boundary value like `session_id` in the signature.
- Initialize `cached = None` before the cache lookup block.
- Add targeted regression tests for the three behaviors above.

Automated hermes-sweeper review.

File: gateway/run.py (hunk 1 above)
Comment (chiefmojo) [reply gateway/run.py:15833]: Fixed in `91765f5c5`. Auto-reset now calls `_clear_auto_reset_session_state()`, which evicts the cached agent for the reused `session_key` and clears session-scoped model/reasoning/pending-note state before `was_auto_reset` is consumed. Added regression coverage in `test_auto_reset_clears_cached_agent_and_session_state`.

**Comment by teknium1** [comment]
Merged via #55595 — your `_cfg_model` normalization fix was cherry-picked onto current main with your authorship preserved in git log (commit d6c53dc).

Thanks for catching the per-turn eviction. Notes on what landed vs. what we dropped:

- **Bug 1 (model normalization mismatch):** merged as-is — this was the real, non-redundant fix. `_cfg_model` is now normalized the same way `AIAgent.__init__` strips it, so vendor-prefixed native-provider config (`deepseek/deepseek-v4-pro`) matches `agent.model` and no longer evicts every turn. Aggregators left untouched.
- **Bug 2 (message_id signature churn):** kept the diagnosis, but reworked the fix to be non-destructive. Deleting the "Triggering message" line removed a real capability (reply/react/pin via the discord tools). Instead we moved the volatile `message_id` out of the cached system prompt and inject it per-turn into the user message, with a static pointer left in the IDs block — preserves both the cache stability *and* the feature. Added a regression test locking the invariant.
- **Auto-reset eviction** (`_clear_auto_reset_session_state` + the auto-reset `_evict_cached_agent`): already on current main via #9893/#48031, so dropped as redundant.
- **`reset_context_note` "scoping bug":** didn't reproduce on current main (the variable is assigned and used in the same scope), so that plumbing was dropped.

Appreciate the contribution.
